kernel/src/grant.rs

```diff
@@ -352,34 +257,30 @@ pub struct Iter<'a, T: 'a + Default> {
         core::slice::Iter<'a, Option<&'static dyn ProcessType>>,
         fn(&Option<&'static dyn ProcessType>) -> Option<&'static dyn ProcessType>,
     >,
+    skips_allowed: bool,
 }
 
-impl<T: Default> Iterator for Iter<'_, T> {
-    type Item = AppliedGrant<T>;
+impl<'a, T: Default> Iterator for Iter<'a, T> {
+    type Item = AppliedGrant<'a, T>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        // Save a local copy of grant_num so we don't have to access `self`
-        // in the closure below.
-        let grant_num = self.grant.grant_num;
-
+        let grant = self.grant;
         // Get the next `AppId` from the kernel processes array that is setup to use this grant.
         // Since the iterator itself is saved calling this function
         // again will start where we left off.
-        let res = self.subiter.find(|process| {
-            // We have found a candidate process that exists in the
-            // processes array. Now we have to check if this grant is setup
-            // for this process. If not, we have to skip it and keep
-            // looking.
-            if let Some(grant_ptr) = process.get_grant_ptr(grant_num) {
-                !grant_ptr.is_null()
-            } else {
-                false
-            }
-        });
-
-        // Check if our find above returned another `AppId`, or if we hit the
-        // end of the iterator. If we found another app, try to access its grant
-        // region.
-        res.map_or(None, |process| self.grant.grant(process.appid()))
+        let skips_allowed = self.skips_allowed;
+        self.subiter.find_map(
+            |process| match AppliedGrant::get_if_allocated(grant, process) {
+                Ok(ag) => Some(ag),
+                Err(None) => None,
+                Err(Some(_err)) => {
+                    if skips_allowed {
+                        None
```

Comment:
```suggestion
                        // This iter was created from `iter_unentered_grants()`,
                        // so we skip entered grants to avoid double entry.
                        None
```